UltraScale Updates

ttown523 opened the pull request:
This pull request updates Tincr export to support UltraScale devices and designs. The device and design export has been testing with RapidSmith2. This PR includes the following:

- For series 7 devices, PAD site names are determined using the Tcl command [get_property NAME [get_package_pins -quiet -of_object $site]]. This assumes there is no more than one package pin per site. This assumption does not hold for ultrascale (see site SYSMONE1_X0Y0 in device xcku025-ffva1156-1-c for an example). The XDLRC generation has been updated to include the real site name (i.e. IOB_X0Y127) instead of the package pin names for ultrascale devices. Series 7 devices still use the package pin naming convention. Also, there was an error where primitive defs were not separated by a new line in the final XDLRC.

- UltraScale devices seem to have VCC/GND "BELs" that don't exist within a primitive site and our just floating in the tile (see image below). 

![image](https://user-images.githubusercontent.com/7906642/26840111-467af974-4aa2-11e7-87b3-7792ba94cef2.png)

This has not been seen in series7 devices and breaks the well-established device hierarchy. So, when 
writing an XDLRC file, these VCC/GND BELs are replaced with a VCC/GND primitive site to preserve the 
device hierarchy and help external tools more easily locate where VCC/GND sources are. 

- The `genFamilyInfo.tcl` file has been completely revamped. The code has been refactored and commented. All unique devices in a family are now searched through when generating the family Info file. Only a few hand edits are required for ultrascale devices.

- `genCellLibrary.tcl` has been updated for ultrascale, and a few bugs have been fixed. This cellLibrary.xml files created from this command do not require any hand edits.

- Net routing export now uses pips instead of wires. This allows external tools to unambiguously reconstruct the physical route of a net 

- RSCPs now have the extension ".rscp" instead of ".tcp"

- Several other code and package refactors

**NOTE: the hand edits to the primitive def files are simply removing the blank line at the end of the file. This is why there are so many files changed**

ttown523 committed:
Ultrascale updates

ttown523 committed:
UltraScale device and design export updates

ttown523 committed:
xdlrc.tcl fix

ttown523 committed:
xdlrc fix

ttown523 committed:
Device file export ultrascale updates

ttown523 committed:
updates

ttown523 committed:
fix vcc gnd site bug

ttown523 committed:
fix vcc and gnd site bug

ttown523 committed:
static primitive site update

ttown523 committed:
ultrascale updates

bradselw commented on tincr/cad/util/util.tcl:
```diff
@@ -1568,6 +1569,16 @@ proc ::tincr::print_object_properties { obj } {
 
+## Removes the speedgrade on the specified part
+#
+# @param partname Full name of the part
```
What happens if the speedgrade is already removed? Ideally, this method would become a no-op.

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -368,15 +368,19 @@ proc ::tincr::write_macros { {filename macros.xml } } {
     # write the XML for new macros
     foreach macro $macros_to_write {
-        tincr::write_macro_xml $macro $xml
+        set internal_nets [tincr::write_macro_xml $macro $xml]
```
Why is a method called `write_macro_nets` returning a list of nets?

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -368,15 +368,19 @@ proc ::tincr::write_macros { {filename macros.xml } } {
+        
     # Create the map of type -> internal netnames for all macros
-    set internal_net_map [dict create]
+    #set internal_net_map [dict create]
```
Remove this line?

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -368,15 +368,19 @@ proc ::tincr::write_macros { {filename macros.xml } } {
+    #set internal_net_map [dict create]
     foreach macro $macros_in_design {
-        dict set internal_net_map [get_property REF_NAME $macro] [get_internal_macro_nets $macro]
+        if { [dict exists $internal_net_map [get_property REF_NAME $macro]] == 0 } {
```
This logic is duplicated throughout Tincr (adding something if it isn't already in a dict). Consider pulling it out into a utility method if it isn't already.

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -393,23 +397,35 @@ proc ::tincr::write_placement_rs2 { {filename placement.rsc} }  {
-                puts $txt "IPROP $cell $property $value"
+    
+    # determine if the current device is a series7 or not
+    set is_series7 [expr {[string first "7" [get_property ARCHITECTURE [get_parts -of [get_design]]]] != -1}]
```
Do we only switch behavior for series 7, or is it series 7 and below?

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -393,23 +397,35 @@ proc ::tincr::write_placement_rs2 { {filename placement.rsc} }  {
+            foreach property [tincr::cells::get_configurable_properties $cell] {
+               
+                set value [get_property $property $cell]
+                # only print the configurations to the file a value exists, and its not the default value
```
"__if__ a value exists"

bradselw replied:
"__it's__ not the default value"

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -457,32 +473,30 @@ proc ::tincr::write_placement_rs2 { {filename placement.rsc} }  {
-    foreach port [get_ports] {
-        if {[get_property PACKAGE_PIN $port] != ""} {
-            puts $txt "PACKAGE_PIN [get_property PACKAGE_PIN $port] [get_ports [get_name $port]]"
+    foreach site [get_sites -of [get_ports]] {
```
`get_sites -quiet`?

bradselw replied:
And below

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -457,32 +473,30 @@ proc ::tincr::write_placement_rs2 { {filename placement.rsc} }  {
+            set net [get_nets -of $bel]
+            if { [llength $net] == 1 } {
+                set port [get_ports -of $net]
+                set bel_name [lindex [split $bel "/"] end]
```
Isn't there a proc in the `bels` ensemble that extracts the name? If not, there should be. This should just call that.

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -457,32 +473,30 @@ proc ::tincr::write_placement_rs2 { {filename placement.rsc} }  {
-    return $sitename
-} 
-
+# @param macro Macro cell instance 
```
More documentation for this proc.

bradselw commented on tincr/io/design/tincr_checkpoints.tcl:
```diff
@@ -550,7 +584,7 @@ proc ::tincr::write_routing_rs2 {args} {
 # @param site_list List of <b>used</b> sites in the design
 # @param channel Output file handle 
-proc write_site_pips { site_list channel } {
+proc write_site_pips { site_list is_series7 channel } {
```
Rather than passing in `is_series7`, the `write_site_pips`, `write_intersite_pins`, and `write_net_routing` procs should compute whether or not the target device is series 7 by calling into a helper proc. IOW move the logic you're using to set `is_series7` into a proc and have all procs call into that.

bradselw commented on tincr/io/device/device_info.tcl:
```diff
@@ -0,0 +1,112 @@
+#   an existing directory, an error will be thrown. 
+# @param partname Full partname of a Vivado device
+proc ::tincr::create_xml_device_info { directory partname } {
+    catch { close_project -quiet  }
```
Would be nice if we had some way of warning the less-informed user that this proc was about to force-close any open projects...

bradselw replied:
Do you have to close the current project for this to work? What about running this in a temporary project?

bradselw commented on tincr/io/device/family_info.tcl:
```diff
@@ -0,0 +1,773 @@
+# 1.) Alternate site pin mappings: When a site type is changed in Vivado to an alternate type, the
+#     pins may be renamed. We cannot determine the which pins are renamed for series7 devices, and
+#     so the user must do this manually. NOT APPICABLE TO ULTRASCALE
+# 2.) Some sites in series 7 return invalid alternate types. Invalid alternate types need to be removed by hand
```
Have you filed a bug with Xilinx about this? Seems like something they would/should be interested in fixing.

bradselw commented on tincr/io/device/family_info.tcl:
```diff
@@ -0,0 +1,773 @@
+# 2.) Some sites in series 7 return invalid alternate types. Invalid alternate types need to be removed by hand
+#     by looking at the alternate types in the device browser GUI
+# 3.) The series7 primitive defs in Tincr contain sites that do not appear in the Tcl interface. At some point, we need
+#     to remove these primitive defs, but for now we can add them to the end of the family info with limited info.
```
Why do we need to remove the primitive defs from Tincr? Is there a better way for getting at intrasite information?

bradselw commented on tincr/io/device/family_info.tcl:
```diff
@@ -0,0 +1,773 @@
+
+# global variable set to 1 if a family is within the series7 of devices
+# Global variables
+set is_series7 0
```
Best to avoid globals if possible. How much time are you really saving by computing this just once as opposed to having each proc compute it?

bradselw commented on tincr/io/device/family_info.tcl:
```diff
@@ -0,0 +1,773 @@
+    set unique_parts [tincr::get_parts_unique $family]
+    set i 1
+    global is_series7
+    set is_series7 [expr {[string first "7" $family] != -1}] 
```
This operation is cheap. I would recommend putting it into an ensemble (e.g. tincr/cad/device/devices.tcl) and just call it from there each time a proc needs to know if the current device is Series7.

bradselw commented on tincr/io/device/family_info.tcl:
```diff
@@ -0,0 +1,773 @@
+    return $compatible_type_map
+}
+
+## Parses the VSRT "addedBels.txt" file for the current architecture,
```
I thought VSRT output primitive definition files? What it this `addedBels.txt` file?

bradselw commented on tincr/io/device/xdlrc.tcl:
```diff
@@ -189,10 +200,19 @@ proc ::tincr::write_xdlrc { args } {
 }
 
-proc ::tincr::write_xdlrc_tile { tile outfile brief } {
+proc ::tincr::write_xdlrc_tile { tile outfile brief is_series7 } {
```
Just compute `is_series7` in the `write_xdlrc_tile` proc. No need to pollute the procedure's prototype.

bradselw commented on tincr/io/device/xdlrc.tcl:
```diff
@@ -718,6 +773,63 @@ proc ::tincr::write_partial_primitive_def { site filename {includeConfigs 0} } {
+    puts $outfile "\t\t\t(conn HARD1VCC P ==> HARD1 HARD1)"
+    puts $outfile "\t\t)"
+    puts -nonewline $outfile "\t)"
+    close $outfile
```
Maybe just check `VCC.def` into Git instead of generating a file every time. File operations are notoriously unreliable, and it's good practice to avoid going out to disk whenever possible. I would put it somewhere under [`cache`](https://github.com/byuccl/tincr/tree/master/cache).

bradselw replied:
Same thing applies for `write_gnd_primitive_def`.

bradselw reviewed (dismissed):
Couple general thoughts:
- This PR is too long. Be nice to your reviewers and consider splitting up the work into separate PRs next time :-)
- Each time you add a proc or edit a proc's documentation, you should re-run Doxygen and update the files under the `docs` folder. Instructions on generating documentation can be found [here](https://github.com/byuccl/tincr/wiki/Development#updating-tincrs-documentation).

ttown523 committed:
Suggested changes on PR#62

ttown523 commented:
Sorry for the long PR. I just finished writing, defending, and editing my thesis so I can finally come up for some air and respond to your requested changes! I have made several changes in the latest push that you can review, and will address your comments in a single comment here so that you are not bombarded with emails. 

> Why is a method called write_macro_xml returning a list of nets?

This function returns a list of internal nets of the macro (i.e. nets that only connect to internal cells of the macro). This information is required later while exporting the routing of a RSCP to export the routing information for those nets. If you don't need these nets, then the return value can be ignored.

> This logic is duplicated throughout Tincr (adding something if it isn't already in a dict). Consider pulling it out into a utility method if it isn't already.

This is a good idea. There are several places in the code that would require updating, so I will look to do this in a future pull request.

> Do we only switch behavior for series 7, or is it series 7 and below?

Vivado does not support Series6 devices and below.

> Isn't there a proc in the bels ensemble that extracts the name? If not, there should be. This should just call that.

No, the `get_name` proc for bels returns the entire name (including the site). There are several places in the code that use this logic, so I will add this function in a future pull request.

> More documentation for this proc.

This proc is documented...I added a @return statement however.

> Rather than passing in is_series7...

I added a function in the `parts` ensemble to test if a device is a series7 device.  I updated the code to use this function instead of passing around the parameter.

> Would be nice if we had some way of warning the less-informed user that this proc was about to force-close any open projects...
Do you have to close the current project for this to work? What about running this in a temporary project?

There is no reason for a user to generate a deviceInfo.xml file when they have an open project, and so there is no reason to support this use case. The user can just open a new instance of Vivado if they need to. I did add a warning that any open projects would be closed in the documentation. 

> Have you filed a bug with Xilinx about this? Seems like something they would/should be interested in fixing.

Fixed in 2017.1

> Why do we need to remove the primitive defs from Tincr? Is there a better way for getting at intrasite information?

The sites that I would want to remove cannot are not represented in the Tcl interface of Vivado. If a user cannot access this information through Tcl, there is no reason that we need the intrasite information in the primitive_defs folder.

>Best to avoid globals if possible. How much time are you really saving by computing this just once as opposed to having each proc compute it?

Removed all globals

> I thought VSRT output primitive definition files? What it this addedBels.txt file?

This file is to make the familyInfo.xml file match the primitive def. In short, GND and VCC BELs are not returned from `get_bels` for UltraScale devices (this may be intentional). In VSRT, you need to add these BELs to create a complete set of connections. VSRT keeps track of the added BELs in an `addedBels.txt` file so the user doesn't have to manually add this information to a family info. Look at the RS2 tech report if you want more details.

> Maybe just check VCC.def into Git instead of generating a file every time. File operations are notoriously unreliable, and it's good practice to avoid going out to disk whenever possible. I would put it somewhere under cache.

I have. This function is here just in case we want to re-generate it or change the format in the future.

> Each time you add a proc or edit a proc's documentation, you should re-run Doxygen and update the files under the docs folder. Instructions on generating documentation can be found here.

I have a few places of documentation to update, so I will re-run Doxygen and create a new pull request after this pull request is accepted.

ttown523 committed:
suppress clock placement messages in cell library generation

ttown523 committed:
zynq updates

ttown523 committed:
zynq updates 2.0

nelsobe reviewed (approved):
Brad - we assume you might be on vacation or something and we need to get this merged while Thomas is still here.  I have reviewed and it looks OK.  Please do take a look at the code and request additional changes if you desire.

bradselw commented:
Sorry about the delay. This looks good to me, thanks Brent for approving!
